tensorboard/data/server/server.rs

```diff
@@ -234,16 +235,114 @@ impl TensorBoardDataProvider for DataProviderHandler {
+                    None => continue,
+                    Some((step, _, _)) => step,
+                };
+                // TODO(@wchargin): Consider tracking this on the time series itself?
+                let max_wall_time = ts
+                    .valid_values()
+                    .map(|(_, wt, _)| wt)
+                    .max()
+                    .expect("have valid values for step but not wall time");
+                run_res.tags.push(data::list_tensors_response::TagEntry {
+                    tag_name: tag.0.clone(),
+                    metadata: Some(data::TensorMetadata {
+                        max_step: max_step.into(),
+                        max_wall_time: max_wall_time.into(),
+                        summary_metadata: Some(*ts.metadata.clone()),
+                        ..Default::default()
+                    }),
+                });
+            }
+            if !run_res.tags.is_empty() {
+                run_res.run_name = run.0.clone();
+                res.runs.push(run_res);
+            }
+        }
+
+        Ok(Response::new(res))
     }
 
     async fn read_tensors(
         &self,
-        _request: Request<data::ReadTensorsRequest>,
+        req: Request<data::ReadTensorsRequest>,
     ) -> Result<Response<data::ReadTensorsResponse>, Status> {
-        Err(Status::unimplemented("not yet implemented"))
+        let req = req.into_inner();
+        let want_plugin = parse_plugin_filter(req.plugin_filter)?;
+        let (run_filter, tag_filter) = parse_rtf(req.run_tag_filter);
+        let num_points = parse_downsample(req.downsample)?;
+        let runs = self.read_runs()?;
+
+        let mut res: data::ReadTensorsResponse = Default::default();
+        for (run, data) in runs.iter() {
+            if !run_filter.want(run) {
+                continue;
+            }
+            let data = data
+                .read()
+                .map_err(|_| Status::internal(format!("failed to read run data for {:?}", run)))?;
+            let mut run_res: data::read_tensors_response::RunEntry = Default::default();
+            for (tag, ts) in &data.tensors {
+                if !tag_filter.want(tag) {
+                    continue;
+                }
+                if plugin_name(&ts.metadata) != Some(&want_plugin) {
+                    continue;
+                }
+
+                let mut points = ts.valid_values().collect::<Vec<_>>();
+                downsample::downsample(&mut points, num_points);
+                let n = points.len();
+                let mut steps = Vec::with_capacity(n);
+                let mut wall_times = Vec::with_capacity(n);
+                let mut values = Vec::with_capacity(n);
+                for (step, wall_time, value) in points {
+                    steps.push(step.into());
+                    wall_times.push(wall_time.into());
+                    // Clone the TensorProto to get a copy to send in the response.
```

Comment:
Cool, this seems fine. If we wanted to be really tight with the locks,
we could change the commit’s tensor values to hold `Arc<TensorProto>`s
and then clone them only after dropping the lock—and maybe one day Prost
and Tonic might [learn about `Arc<_>`s][arc] and remove the need to clone
at all. But certainly no action required.

[arc]: https://github.com/danburkert/prost/issues/429

Reply:
Yeah, I realize it's not exactly ideal to be cloning this here, but SGTM to defer further optimization for now.  I'd found myself wondering whether TensorProto was even the right type for the commit to hold in the first place (since we're forcing it into an ndarray ultimately, it could be slimmed down quite a bit, and maybe made more ergonomic to manipulate).  If we revisit that, we could also consider reference counting to avoid the clone-inside-critical-section here.